pallets/elections/src/impls.rs

```diff
@@ -0,0 +1,316 @@
+use crate::{Config, ErasReserved, MembersPerSession, Pallet, SessionValidatorBlockCount};
+use frame_election_provider_support::sp_arithmetic::Perquintill;
+use frame_support::{pallet_prelude::Get, traits::Currency};
+use primitives::TOKEN;
+use sp_staking::{EraIndex, SessionIndex};
+use sp_std::vec::Vec;
+
+fn scale_total_exposure(total: u128) -> u128 {
+    // to cover minimal possible theoretical stake (25k) and avoid loss of accuracy we need to scale
+    total / TOKEN
```

Comment:
This scaling can be improved. If for some reason one would run it with some strange token distribution on a dev chain, this has the potential to panic (at the place where we cast to u32). And in general this is not quite robust.

I think the correct way to do that would be to make the total (sum over validators) maximum possible reward per era equal to u32::MAX. And then scale according to that. This will give you the optimal precision, and will never lead to unexpected behavior...

Reply:
u32::MAX can be replaced by some large round number, if we prefer that, like 10**9.

Reply:
Done

Reply:
This still looks similar as before, only some code was moved. 
What you want to avoid is this kind of random scaling `(exposure.total.into() / TOKEN) as u32` because, why would you scale by `TOKEN`? That's a pretty random constant in this context. 

Let `v1, ..., vN` be the validators for this era. Let `e1, e2, ..., eN` be their exposure (as u128) and let `e=e1+e2+...+eN`. You want the scaling `(e1', e2', ..., eN')` (as `u32`) to be such that `e' = e1'+e2'+...+eN'` \approx `MAX_REWARD` (you will probably not get equality because of rounding errors, but make sure we always round down). Now, for validator `vi` the target reward for a single session should be `ei'/sessions_per_era`, i.e.:
-  if it is not in the committee, or it created >=90% fraction of blocks it was supposed to create, then it would get `ei'/sessions_per_era` reward
- otherwise `<ei'/sessions_per_era` proportionally to how many blocks it created.

Reply:
Ok, now it should be as you described. Also added lenient uptime in this pr

Reply:
Great, this looks good now. But you don't need to consider the case `sum_totals <= MAX_REWARD` separately. Just use this formula for scaling:
`(t*MAX_REWARD)/sum_totals` 
If you are scared that `t*MAX_REWARD` might overflow `u128` then you can either:
1. ignore it, because it will not,
2. do the multiplication in U256,
3. use `saturating_mul`

I think 3. is the cleanest of them.

Reply:
done, I applied the option 3.